### PR TITLE
refactor: Add separate ECD and GBM defaults schemas

### PR DESCRIPTION
**Pull request by ksbrar**
~~The last test failure I'm seeing is in `test_old_models::test_predict_deprecated_model`. It occurs because the `gbm_adult_census_income_v061` model was trained when we had a common defaults schema so the model [hyperparameters](https://predibase-public-us-west-2.s3.us-west-2.amazonaws.com/ludwig_unit_tests/gbm_adult_census_income_v061.zip) file includes the full ECD defaults schema.~~

~~It looks like loading this old model bypasses the config transformations so I'm not sure what the best way is to resolve the failure. CC @jppgks @justinxzhao for ideas?~~ Nevermind I just messed up my function.

### CODE REVIEW

File: ludwig/schema/defaults/ecd.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    VECTOR,
+)
+from ludwig.schema import utils as schema_utils
+from ludwig.schema.defaults.base import BaseDefaultsConfig
+from ludwig.schema.defaults.utils import DefaultsDataclassField
+from ludwig.schema.features.base import BaseFeatureConfig
+
+
+@DeveloperAPI
+@dataclass
```
Comment (tgaddair) [inline ludwig/schema/defaults/ecd.py:26]: `ludwig_dataclass`?

File: ludwig/schema/defaults/gbm.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from ludwig.api_annotations import DeveloperAPI
+from ludwig.constants import BINARY, CATEGORY, NUMBER
+from ludwig.schema import utils as schema_utils
+from ludwig.schema.defaults.base import BaseDefaultsConfig
+from ludwig.schema.defaults.utils import DefaultsDataclassField
+from ludwig.schema.features.base import BaseFeatureConfig
+
+
+@DeveloperAPI
+@dataclass
```
Comment (tgaddair) [inline ludwig/schema/defaults/gbm.py:12]: `ludwig_dataclass`?

**Review by tgaddair — APPROVED** [review]
Nice change! Couple minor nits.

**Comment by tgaddair** [comment]
@ksbrar the way to resolve the test failure is to add a backwards compatibility hook that removes the offending default values.

File: ludwig/schema/defaults/ecd.py (hunk 1 above)
Comment (ksbrar) [reply ludwig/schema/defaults/ecd.py:26]: fixed

File: ludwig/schema/defaults/gbm.py (hunk 2 above)
Comment (ksbrar) [reply ludwig/schema/defaults/gbm.py:12]: fixed

File: ludwig/utils/backward_compatibility.py [hunk 3]
```diff
@@ -798,6 +800,17 @@ def upgrade_missing_hyperopt(config: ModelConfigDict) -> ModelConfigDict:
     return config
 
 
+@register_config_transformation("0.7", ["defaults"])
```
Comment (ksbrar) [inline ludwig/utils/backward_compatibility.py:803]: @tgaddair @justinxzhao What version should I put here?

**Review by justinxzhao — APPROVED** [review]
Nice!

**Review by arnavgarg1 — APPROVED** [review]
Looks great!

Just wondering - should we add an equivalent test like `test_ecd_defaults_schema` for GBMs as well?

File: tests/ludwig/config_validation/test_validate_config_misc.py [hunk 4]
```diff
@@ -266,8 +266,8 @@ def test_validate_with_preprocessing_defaults():
     check_schema(config)
 
 
-def test_defaults_schema():
-    schema = DefaultsConfig()
+def test_ecd_defaults_schema():
```
[the remark this answers is not in the document]
Comment (jeffkinnison) [reply tests/ludwig/config_validation/test_validate_config_misc.py:269]: Ah, looks like @arnavgarg1 beat me to the punch on that one!

**Review by connor-mccorm — APPROVED** [review]

File: tests/ludwig/config_validation/test_validate_config_misc.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ksbrar) [reply tests/ludwig/config_validation/test_validate_config_misc.py:247]: added

File: tests/ludwig/config_validation/test_validate_config_misc.py [hunk 5]
```diff
@@ -277,6 +255,13 @@ def test_defaults_schema():
     assert LOSS in schema.category.to_dict()
 
 
+def test_gbm_defaults_schema():
+    schema = GBMDefaultsConfig()
+    assert AUDIO not in schema.to_dict()
+    assert schema.category.encoder.dropout == 0.0
```
Comment (ksbrar) [inline tests/ludwig/config_validation/test_validate_config_misc.py:261]: @jeffkinnison @arnavgarg1 It just occurred to me - do we actually support encoders/decoders for GBM?

**Pull request merged**